### PR TITLE
Trust all the trust anchors in the keystore

### PR DESCRIPTION
**Pull request by mcpherrinm**
This change trusts everything in a keystore when no trustAlias is specified.

Without this change, it is difficult to do a hitless replacement of a root CA, as the truststore used by the jdbc driver must change at the same time as the vtgate's server certificate.  That's a feat tricky to do hitlessly at scale.

The current behavior of only trusting the first entry in the truststore is unexpected and unusual.  While it is documented, I have not seen other pieces of software doing that.  The Go portions of vitess, for example, load all entries from a PEM ca file,
https://github.com/vitessio/vitess/blob/fd262fe98ab30533bffb6510a0154825ea94ecb3/go/vt/vttls/vttls.go#L69-L78

I haven't updated tests or documentation yet, and am really just looking for feedback on the approach first. 

Risk:  This shouldn't break any existing user, as the first certificate would be a member of all the certificates, but it could potentially cause a user who was relying on only the first entry to be trusted to trust additional roots they were not expecting.  That seems unlikely to me, and those users can use the TrustedAlias to specify which root they intend to trust in that case.

### CODE REVIEW

File: java/grpc-client/src/main/java/io/vitess/client/grpc/GrpcClientFactory.java [hunk 1]
```diff
@@ -268,22 +272,18 @@ private KeyStore loadKeyStore(final File keyStoreFile, String keyStorePassword)
...
-    } catch (KeyStoreException exc) {
-      return null;
-    }
-
-    while (aliases.hasMoreElements()) {
-      final String alias = aliases.nextElement();
-      final X509Certificate[] certCollection = loadCertCollectionForAlias(keyStore, alias);
-      if (certCollection != null) {
-        return certCollection;
+      PKIXParameters p = new PKIXParameters(keyStore);
```
Comment (mpawliszyn) [inline java/grpc-client/src/main/java/io/vitess/client/grpc/GrpcClientFactory.java]: Checkstyle is complaining how about `param`

File: java/grpc-client/src/main/java/io/vitess/client/grpc/GrpcClientFactory.java [hunk 2]
```diff
@@ -268,22 +272,18 @@ private KeyStore loadKeyStore(final File keyStoreFile, String keyStorePassword)
...
+      X509Certificate[] certificates = p.getTrustAnchors().stream()
+          .map(TrustAnchor::getTrustedCert)
+          .toArray(X509Certificate[]::new);
+      if (certificates.length == 0) {
+        return null;
       }
+      return certificates;
+    }  catch (InvalidAlgorithmParameterException | KeyStoreException e) {
```
Comment (mpawliszyn) [inline java/grpc-client/src/main/java/io/vitess/client/grpc/GrpcClientFactory.java]: Checkstyle is also complaining here how about `exc`

Comment (mcpherrinm) [reply java/grpc-client/src/main/java/io/vitess/client/grpc/GrpcClientFactory.java]: Fixed the checkstyle issues, thanks.

**Comment by sougou** [comment]
Note that tests are also failing.

**Comment by leoxlin** [comment]
Please rebase and address the dependency check issue:

```
[WARNING] Unused declared dependencies found:
[WARNING]    io.grpc:grpc-netty:jar:1.16.0:compile
[WARNING]    io.netty:netty-handler:jar:4.1.30.Final:compile
[WARNING]    io.opentracing.contrib:opentracing-grpc:jar:0.0.12:compile
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Vitess Java Client libraries and Hadoop support [Parent] SUCCESS [  2.764 s]
[INFO] vitess-client ...................................... SUCCESS [ 26.219 s]
[INFO] vitess-grpc-client ................................. FAILURE [ 27.861 s]
[INFO] vitess-jdbc ........................................ SKIPPED
[INFO] vitess-example ..................................... SKIPPED
[INFO] vitess-hadoop ...................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 57.372 s
[INFO] Finished at: 2019-07-22T20:44:47+00:00
[INFO] Final Memory: 48M/772M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:2.8:analyze-only (analyze) on project vitess-grpc-client: Dependency problems found -> [Help 1]
```

**Comment by sougou** [comment]
@mcpherrinm Do you want to resurrect this? If so, tests need fixing.

**Comment by mcpherrinm** [comment]
I'm not able to work on this at the moment; I've reassigned the ticket internally to another team here but I'm not sure when they'll be able to fix this.   We don't have another root rotation coming up for a while so it's unlikely we'll have another outage soon, so it's not a very high priority.

**Comment by mcpherrinm** [comment]
It doesn't look like anyone else picked this up, and I'm no longer working somewhere that uses Vitess, so I'm going to close it.

**Pull request closed without merging**